### ISSUE TITLE
axes.format_{x,y} data should have a setter

### ISSUE THREAD

**Issue by story645** [matplotlib/matplotlib#17286]
In the [Date tick labels](https://matplotlib.org/3.1.1/gallery/text_labels_and_annotations/date.html#sphx-glr-gallery-text-labels-and-annotations-date-py) example, the code for setting the cursor coords message overwrites the default formatters:
```python
ax.format_xdata = mdates.DateFormatter('%Y-%m-%d')
ax.format_ydata = lambda x: '$%1.2f' % x  # format the price.
```
This creates confusion because runs counter to how basically everything else in matplotlib is set via setter but especially the formatters:
```python
ax.xaxis.set_major_formatter(mdates.DateFormatter('%Y'))
```
and [the docs](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.axes.Axes.format_xdata.html?highlight=format_xdata#matplotlib.axes.Axes.format_xdata) aren't especially clear either on this being a property:
```
Axes.format_xdata(self, x)

    Return x formatted as an x-value.

    This function will use the fmt_xdata attribute if it is not None, else will fall back on the xaxis major formatter.
```
To be consistent with the API, I'm proposing properties/functions of the form:
* `ax.{x,y}data_formatter(mdates.DateFormatter('%Y-%m-%d'))`
* `ax.data_formatter(which=(x,y, both), mdates.DateFormatter('%Y-%m-%d'))`

**Comment by anntzer** [comment]
Really this should be called something like `set_xcursor_formatter`, `data` is just too vague.

**Comment by story645** [comment]
Agree, and should also steal from #16715 that it should take a str, callable, or mticker instance.

**Comment by story645** [comment]
And xref #9593 'cause fixing this issue should maybe help fix that one?
